### PR TITLE
change visibility of groth16 proof

### PR DESCRIPTION
**Pull request by wz14**
This PR changes the visibility of some structures and functions of groth16.
It's helpful to extract the *real* groth16 proof from r0 proof receipt for those teams writing on-chain verifier themselves.

### CODE REVIEW

**Comment by nategraf** [comment]
Hi @wz14! These fields and functions are not `pub` because they are not part of the stable API surface. We actually have plans to change the underlying implementation of these types from Arkworks to another implementation. If they were `pub`, this would be a breaking change.

In order to extract the Groth16 proof from its encoded form as the seal. This is what we do in `risc0-ethereum` [1].

[1] https://github.com/risc0/risc0-ethereum/blob/ac4a73fbed87ebad34306e7808fd951dbfae3f75/contracts/src/groth16/RiscZeroGroth16Verifier.sol#L163

**Comment by flaub** [comment]
@nategraf could we make them `pub` but also mark as `stability::unstable`? That way we can unblock today, with no guarantee that these functions will exist in a future release.

**Comment by nategraf** [comment]
@flaub, sure! That seems like a good way to handle this. @wz14, does that work for you? Here is an example of what we are referring to. https://github.com/risc0/risc0/blob/a917620feaa03b030db8e5f6d38d7c8b764cdf99/risc0/zkvm/src/guest/env/mod.rs#L444

If you mark these are unstable, I'll re-review this PR and would be happy to approve it.

**Comment by wz14** [comment]
@nategraf Hi, I've marked `unstable` to these functions. Please help review this PR.

File: risc0/groth16/src/verifier.rs
Comment (flaub) [inline risc0/groth16/src/verifier.rs:161]:
```suggestion
pub struct Fr(#[serde(with = "serde_ark")] ark_bn254::Fr);
```

This needs to remain private since `ark_fr` provides access in an unstable way.

File: risc0/groth16/src/verifier.rs
Comment (flaub) [inline risc0/groth16/src/verifier.rs:188]: ditto here

**Comment by wz14** [comment]
Thanks. @flaub Already Corrected.

**Review by nategraf — APPROVED** [review]
Looks good! Thanks for making those changes

**Comment by wz14** [comment]
Hi, here seems to need a Vercel approval from you @nategraf

**Comment by nategraf** [comment]
> Hi, here seems to need a Vercel approval from you @nategraf 

Vercel is for previewing the website. You shouldn't need it here. Merging is currently blocked because the branch is out of date with `main`. If you merge `main`, it should be unblocked. Let me know you need another review at that point

**Comment by wz14** [comment]
> > Hi, here seems to need a Vercel approval from you @nategraf
> 
> Vercel is for previewing the website. You shouldn't need it here. Merging is currently blocked because the branch is out of date with `main`. If you merge `main`, it should be unblocked. Let me know you need another review at that point

Already catch up to `main` branch. Seems still to be blocked.

**Comment by nategraf** [comment]
Ah, I see now. `cargo sort --check` us failing. Let me know when you've resolved that. Alternatively I could push to your branch to resolve minor things. Whichever works for you.
https://github.com/risc0/risc0/actions/runs/11210344610/job/31164655768#step:11:1

**Comment by wz14** [comment]
> Ah, I see now. `cargo sort --check` us failing. Let me know when you've resolved that. Alternatively I could push to your branch to resolve minor things. Whichever works for you. https://github.com/risc0/risc0/actions/runs/11210344610/job/31164655768#step:11:1

Really thx. I'm guessing the problem is the dependencies for groth16 are not sorted. It should be passed now. @nategraf  Need another approval to run the CI.

**Comment by wz14** [comment]
Hi, update Cargo.lock for benchmark and examples. Need another approval to run CI. @nategraf

**Review by nategraf — APPROVED** [review]

**Pull request merged**
